examples/chain-of-density/chain_of_density.py

```diff
@@ -0,0 +1,223 @@
+                {
+                    "role": "user",
+                    "content": f"Here is the first summary : {original_summary}",
+                },
+                {
+                    "role": "user",
+                    "content": f"Here is the second summary : {self.summary}",
+                },
+            ],
+            response_model=MaybeOmittedEntities,
+        )
+        if evaluation.omitted_entities:
+            omitted_entities = [
+                entity.entity_name for entity in evaluation.omitted_entities
+            ]
+            raise ValueError(
+                f"The following entities were omitted from the summary - {','.join(omitted_entities)}. Please generate a new summary which does not omit them."
+            )
+
+        return self
+
+
+def rewrite_summary(
+    article: str,
+    existing_summary: str,
+    entity_ctx: str,
+    error_msgs: List[str] = [],
+    remaining_retries=3,
+):
+    # # We then perform a new summary and validate that the entity density has increased ( We have not lost any entities )
+    completion = openai.ChatCompletion.create(
+        model="gpt-3.5-turbo-16k",
+        functions=[RewrittenSummary.openai_schema],
+        function_call={"name": RewrittenSummary.openai_schema["name"]},
+        max_retries=2,
+        messages=[
+            {
+                "role": "system",
+                "content": "You are about to be given an article, an existing summary of the article and some new entities. Please use the information to rewrite the summary to make it denser and more concise so that it covers every entity and detail from the previous summary plus the missing entities",
+            },
+            {"role": "user", "content": f"Here is the article : {article}"},
+            {
+                "role": "user",
+                "content": f"Here is the most recent article : {existing_summary}",
+            },
+            {
+                "role": "user",
+                "content": f"Here is some information on entities you should include in the rewritten summary: {entity_ctx}",
+            },
+            *error_msgs,
+        ],
+    )
+    try:
+        new_summary = RewrittenSummary.from_response(
+            completion, validation_context={"prev_summary": existing_summary}
+        )
+        return new_summary
+    except (ValidationError, JSONDecodeError) as e:
+        if remaining_retries == 0:
+            raise e
+        error_msgs = []
+        error_msgs.append(dict(**completion.choices[0].message))
+        error_msgs.append(
+            {
+                "role": "user",
+                "content": f"Recall the function correctly, exceptions found\n{e}",
+            }
+        )
+        return rewrite_summary(
+            article,
+            existing_summary,
+            entity_ctx,
+            error_msgs,
+            remaining_retries=remaining_retries - 1,
+        )
+
+
+def summarize_article(article: str, summary_steps: int = 3):
```

Comment:
Think this and rewrite summary could be one function.

Reply:
Hmm was originally using the model_validate() function which required the use of a context. This didn't have a native retry from what I could see since I had to parse it using `new_summary = RewrittenSummary.from_response(..code)` and so I wrote up a new function for it.

Will simplify the logic and move away from using context validation here.